pkg/imageverifier/bindir/bindir.go

```diff
@@ -0,0 +1,189 @@
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package bindir
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/pkg/imageverifier"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+const mib = 1 << 20 // Number of bytes in 1 MiB
+
+type Config struct {
+	BinDir             string        `toml:"bin_dir"`
+	MaxVerifiers       int           `toml:"max_verifiers"`
+	PerVerifierTimeout time.Duration `toml:"per_verifier_timeout"`
+}
+
+type ImageVerifier struct {
+	config *Config
+}
+
+var _ imageverifier.ImageVerifier = (*ImageVerifier)(nil)
+
+func NewImageVerifier(c *Config) *ImageVerifier {
+	return &ImageVerifier{
+		config: c,
+	}
+}
+
+func (v *ImageVerifier) VerifyImage(ctx context.Context, name string, desc ocispec.Descriptor) imageverifier.Judgement {
+	entries, err := os.ReadDir(v.config.BinDir)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return imageverifier.Judgement{
+				OK:     true,
+				Reason: fmt.Sprintf("image verifier directory %v does not exist", v.config.BinDir),
+			}
+		}
+
+		return imageverifier.Judgement{
+			OK:     false,
+			Reason: fmt.Sprintf("failed to list directory contents: %v", err),
+		}
+	}
+
+	if len(entries) == 0 {
+		return imageverifier.Judgement{
+			OK:     true,
+			Reason: fmt.Sprintf("no image verifier binaries found in %v", v.config.BinDir),
+		}
+	}
+
+	reason := &strings.Builder{}
+	for i, entry := range entries {
+		if (i+1) > v.config.MaxVerifiers && v.config.MaxVerifiers >= 0 {
+			log.G(ctx).Warnf("image verifiers are being skipped since directory %v has %v entries, more than configured max of %v verifiers", v.config.BinDir, len(entries), v.config.MaxVerifiers)
+			break
+		}
+
+		bin := entry.Name()
+		msg, err := v.runVerifier(ctx, bin, name, desc)
+		if err != nil {
+			if ee := (&exec.ExitError{}); errors.As(err, &ee) && ee.ProcessState.Exited() {
+				ec := ee.ProcessState.ExitCode()
+				return imageverifier.Judgement{
+					OK:     false,
+					Reason: fmt.Sprintf("verifier %v rejected image (exit code %v): %v", bin, ec, msg),
+				}
+			}
+			return imageverifier.Judgement{
+				OK:     false,
+				Reason: fmt.Sprintf("verifier %v rejected image due to error: %v", bin, err),
+			}
+		}
+
+		if i > 0 {
+			reason.WriteString(", ")
+		}
+		reason.WriteString(fmt.Sprintf("%v => %v", v, msg))
+	}
+
+	return imageverifier.Judgement{
+		OK:     true,
+		Reason: reason.String(),
+	}
+}
+
+func (v *ImageVerifier) runVerifier(ctx context.Context, bin string, imageName string, desc ocispec.Descriptor) (msg string, err error) {
+	ctx, cancel := context.WithTimeout(ctx, v.config.PerVerifierTimeout)
+	defer cancel()
+
+	binPath := path.Join(v.config.BinDir, bin)
+	args := []string{
+		"-name", imageName,
+		"-digest", desc.Digest.String(),
```

Comment:
Looking for feedback: should we include other descriptor fields? There is a lot of data in there, but I think most verifiers would just use the digest.

Reply:
I think most of the fields would be useful, with the exception of `URLs` and `Data`.  `Size` is maybe another questionable one since that'd just be the size of the index or manifest itself.

* `MediaType` and `ArtifactType` - both help determine the content of the thing being pulled, and a verifier might want to limit that type of content
* `Annotations` - Could be used to allow or deny some images based on how they were built, such as if an image was generated from a CI system.  Combined with a signature over the manifest, an annotation could be attested data.
* `Platform` - Could be used as an easy way to restrict containerd's automatic platform fallback for Arm versions, restrict Linux images from being run on FreeBSD, etc.

Reply:
I will add those. 

> Could be used as an easy way to restrict containerd's automatic platform fallback for Arm versions, restrict Linux images from being run on FreeBSD, etc.
 
Aside: I've personally seen a case where the automatic fallback from `amd64` to `i386` causes a huge performance degradation and it would be useful to be able to restrict that behavior via this mechanism.

Reply:
What are your thoughts on passing the descriptor via stdin in JSON format? This makes it difficult to evolve the API, since stdin will be pinned to a specific meaning, but some of the fields are awkward to pass via flags, like a string map for annotations.

Reply:
stdin as a JSON object makes sense to me, though we could wrap it in some other object as future-proofing for more data to be passed in the future.

@dmcgowan I'd be interested in your thoughts here.

Reply:
stdin makes sense to me as well. Ideally set a `--stdin-media-type` or something though to type the input object. Creating an OCI index as the input could make sense here as well.

Reply:
Passing a Descriptor over stdin is implemented in e87da1322a70e0039f1964c5669e0031bf60b0eb